### PR TITLE
e2e: run only [Happy] tests on PRs, full suite nightly. Closes #960

### LINKED ISSUE #960: Add e2e test tiers to reduce PR CI time
## Problem

E2E test execution time is growing as we add more tests. Every PR runs the full suite, slowing down the feedback loop for contributors.

## Proposal

Introduce two e2e test tiers:

**Tier 1 — PR gate (required, every PR)**
- Core routing and broker functionality
- Session management basics
- Critical auth paths
- Tool listing and tool/call happy paths

**Tier 2 — Nightly/opt-in (scheduled, or manually triggered)**
- Edge cases and error handling
- Performance/slow tests
- External service integration
- Full cross-cutting scenarios (e.g. multi-server federation, custom paths)

## Implementation ideas

- Use Ginkgo labels (e.g. `Label("tier1")`, `Label("tier2")`) to tag tests
- Tier 1 runs via `make test-e2e` (default, used in PR CI)
- Tier 2 runs via `make test-e2e-full` or a label filter override
- Add a nightly GitHub Actions workflow for the full suite
- Add a `/test-e2e-full` comment trigger for opt-in full runs on PRs

## Acceptance criteria

- [ ] Tests are labeled with tier annotations
- [ ] Tier 1 suite identified and kept under a target time (e.g. 5 min)
- [ ] PR CI only runs tier 1
- [ ] Nightly workflow runs both tiers
- [ ] Docs updated with guidance on which tier new tests belong in

### PR DESCRIPTION
**Pull request by manik3160**
The e2e suite now takes long enough that running everything on each PR
creates unnecessary friction. Tests are already labelled with [Happy],
[Full], [multi-gateway], [Elicitation], and [Auth] in their names, so
the split is straightforward.

- `test-e2e-ci` now focuses on `[Happy]` tests only — 20 tests, the
  core routing, session, tool, and registration scenarios that cover
  most regressions quickly.
- `test-e2e-ci-full` is a new target that runs everything, matching
  what `test-e2e-ci` used to do.
- The on-demand `/test-e2e full` comment trigger is updated to call
  `test-e2e-ci-full` so it still runs the complete suite.
- A new nightly workflow runs `test-e2e-ci-full` on main at 02:00 UTC,
  catching any regressions in the slower [Full] and [multi-gateway]
  tests without blocking PR feedback.

No test files changed. No labels added or removed. The only change to
`e2e.yaml` is indirect — it calls `make test-e2e-ci` which now runs
tier 1 only.

Closes #960 

## Summary by CodeRabbit

* **Chores**
  * Added nightly automated end-to-end test runs with per-ref concurrency and manual trigger; improved failure log collection and updated on-demand E2E workflow.
  * Toolchain setup for E2E workflows now reads the Go version from project metadata.

* **Tests**
  * Added CI targets for full-suite nightly runs and a focused "happy-path" run; PR-gate uses a faster subset with fail-fast.

* **Documentation**
  * Documented test tiers and how specs are selected for PR vs nightly runs.

### CODE REVIEW

File: .github/workflows/e2e-nightly.yaml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    timeout-minutes: 60
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Setup Go
+        uses: actions/setup-go@v5
+        with:
+          go-version: '1.24'
```
[the remark this answers is not in the document]
Comment (manik3160) [reply .github/workflows/e2e-nightly.yaml]: Done

[the remark this answers is not in the document]
Comment (manik3160) [reply .github/workflows/e2e-nightly.yaml]: Done

File: build/e2e.mk [hunk 2]
```diff
@@ -38,7 +38,11 @@ test-e2e-watch: test-e2e-deps ## Run e2e tests in watch mode for development
 
 # CI-specific target that assumes cluster exists
 .PHONY: test-e2e-ci
-test-e2e-ci: test-e2e-deps enable-debug-logging ## Run e2e tests in CI (no setup, fail fast)
+test-e2e-ci: test-e2e-deps enable-debug-logging ## Run tier 1 e2e tests in CI ([Happy] only, fail fast)
+	$(GINKGO) -v --tags=e2e --timeout=$(E2E_TIMEOUT) --fail-fast --focus="\[Happy\]" ./tests/e2e
```
[the remark this answers is not in the document]
Comment (manik3160) [reply build/e2e.mk]: Implemented

**Comment by manik3160** [comment]
This PR is ready to review now !

**Comment by david-martin** [comment]
Hey, thanks for the contribution. Just a heads up: reviews are currently being prioritised around the v0.7 milestone. PRs tied to v0.7 issues will be reviewed first. We'll get to this one, but it may take a bit longer than usual.

File: build/e2e.mk [hunk 3]
```diff
@@ -3,6 +3,8 @@
 GINKGO = $(shell pwd)/bin/ginkgo
 GINKGO_VERSION = v2.27.2
 E2E_TIMEOUT ?=15m
+# Tier-1 PR gate: single --focus regex for specs tagged in titles with [Happy]
+E2E_GINKGO_FOCUS_TIER1 = --focus="\[Happy\]"
```
Comment (jasonmadigan) [inline build/e2e.mk]: **Blocker: regex bug excludes `[Happy,URLElicitation]` tests.**

Ginkgo uses Go's RE2 regex engine. The pattern `\[Happy\]` requires `]` immediately after `Happy`, so it matches `[Happy]` but NOT `[Happy,URLElicitation]`. There are 4 specs in `url_elicitation_test.go` (lines 60, 96, 265, 377) tagged `[Happy,URLElicitation]` that intend to be tier 1 but silently won't match.

Fix: change to `\[Happy` (drop the trailing `\]`) or retag those specs to use separate brackets `[Happy] [URLElicitation]`.

File: build/e2e.mk [hunk 4]
```diff
@@ -38,7 +40,11 @@ test-e2e-watch: test-e2e-deps ## Run e2e tests in watch mode for development
 
 # CI-specific target that assumes cluster exists
 .PHONY: test-e2e-ci
-test-e2e-ci: test-e2e-deps enable-debug-logging ## Run e2e tests in CI (no setup, fail fast)
+test-e2e-ci: test-e2e-deps enable-debug-logging ## Run tier 1 e2e tests in CI ([Happy] only, fail fast)
+	$(GINKGO) -v --tags=e2e --timeout=$(E2E_TIMEOUT) --fail-fast $(E2E_GINKGO_FOCUS_TIER1) ./tests/e2e
```
Comment (jasonmadigan) [inline build/e2e.mk]: **Blocker: untagged tests default to nightly-only.**

Only 30 of 72 specs have a matchable `[Happy]` tag. The remaining 42 (including all 18 in `tool_discovery_test.go`) silently fall to nightly-only. The safe default should be inverted: untagged tests should run on PRs, not be excluded.

Options:
- Tag the remaining tests explicitly (`[Happy]` or `[Extended]`)
- Add a CI lint step that fails when a spec has no tier tag
- Change the approach so `--skip` excludes known-slow tests instead of `--focus` including known-fast ones

File: .github/workflows/e2e-nightly.yaml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    timeout-minutes: 60
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Setup Go
+        uses: actions/setup-go@v5
+        with:
+          go-version-file: go.mod
```
Comment (jasonmadigan) [inline .github/workflows/e2e-nightly.yaml:24]: Nit: this uses `go-version-file: go.mod` while the existing `e2e.yaml` and `e2e-on-demand.yaml` hardcode `go-version: '1.24'`. The `go.mod` approach is better but they'll diverge on the next Go bump. Worth aligning all e2e workflows to one approach.

File: tests/e2e/CLAUDE.md [hunk 6]
```diff
@@ -1,5 +1,9 @@
 # E2E Tests
 
+## Test tiers (PR CI vs nightly)
+
+Tier selection uses bracket tags in Ginkgo `Describe` / `It` titles (for example `[Happy]`, `[Full]`, `[multi-gateway]`). PR CI runs only `[Happy]` specs via `make test-e2e-ci` (`E2E_GINKGO_FOCUS_TIER1` in `build/e2e.mk`). The full suite, including other tags, runs under `make test-e2e-ci-full` (nightly workflow and on-demand full runs). When you add a spec, put it in tier 1 only if it is a fast happy-path regression check; otherwise use a non-`[Happy]` tag so it stays in the nightly or opt-in path.
+
```
Comment (jasonmadigan) [inline tests/e2e/CLAUDE.md:17]: Nit: `[Negative]` is used in `custom_tls_test.go` but isn't listed in this tag scheme. Worth adding for completeness.

**Review by jasonmadigan — CHANGES_REQUESTED** [review]
Two issues need fixing before this can merge, plus a few smaller items.

The approach is sound and well-motivated -- splitting e2e into PR-fast and nightly-full is the right call. The implementation has a regex bug that silently excludes intended tier-1 tests, and the default for untagged tests is inverted (untagged = nightly-only, should be untagged = runs on PRs).

**Comment by jasonmadigan** [comment]
needs a rebase from `main`

File: .github/workflows/e2e-nightly.yaml [hunk 7]
```diff
@@ -0,0 +1,39 @@
+name: E2E Tests (nightly)
+
+on:
+  schedule:
+    - cron: '0 2 * * *'
+  workflow_dispatch:
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
```
[the remark this answers is not in the document]
Comment (manik3160) [reply .github/workflows/e2e-nightly.yaml:10]: This workflow only runs on schedule/dispatch, never on PRs, so github.event.pull_request.number would always be empty. The existing expression is correct.

File: .github/workflows/e2e-nightly.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (manik3160) [reply .github/workflows/e2e-nightly.yaml:24]: Matches the existing convention used in all other e2e workflows in this repo. Pinning SHA hashes would be a repo-wide change, out of scope here.

**Comment by manik3160** [comment]
Hey @jasonmadigan ! Thanks for the feedback. I've addressed the review comments and rebased onto main :)

File: build/e2e.mk [hunk 8]
```diff
@@ -38,7 +42,11 @@ test-e2e-watch: test-e2e-deps ## Run e2e tests in watch mode for development
 
 # CI-specific target that assumes cluster exists
 .PHONY: test-e2e-ci
-test-e2e-ci: test-e2e-deps enable-debug-logging ## Run e2e tests in CI (no setup, fail fast)
+test-e2e-ci: test-e2e-deps enable-debug-logging ## Run PR-gate e2e tests in CI (excludes slow tier 2 suites, fail fast)
+	$(GINKGO) -v --tags=e2e --timeout=$(E2E_TIMEOUT) --fail-fast $(E2E_GINKGO_SKIP_TIER2) ./tests/e2e
+
+.PHONY: test-e2e-ci-full
+test-e2e-ci-full: test-e2e-deps enable-debug-logging ## Run all e2e tests in CI (tier 1 + 2, fail fast)
```
Comment (Patryk-Stefanski) [inline build/e2e.mk]: Nit: `test-e2e-ci-full` uses `--fail-fast`, which means the nightly run stops at the first failure. For a nightly regression sweep you typically want the full picture — all failures, not just the first. Consider dropping `--fail-fast` from the full target so nightly runs report every broken spec.

Not a blocker, but worth considering.

**Review by Patryk-Stefanski — COMMENTED** [review]
## Review

All previous review comments have been addressed. Two minor observations inline — neither is a blocker.

File: build/e2e.mk (hunk 8 above)
Comment (manik3160) [reply build/e2e.mk]: Good catch ! Thanks :)

**Review by jasonmadigan — APPROVED** [review]
